File: app/src/main/java/com/kickstarter/services/interceptors/ApiRequestInterceptor.java
package com.kickstarter.services.interceptors;

import android.net.Uri;

import com.google.firebase.iid.FirebaseInstanceId;
import com.kickstarter.libs.Build;
import com.kickstarter.libs.CurrentUserType;
import com.kickstarter.libs.perimeterx.PerimeterXClientType;
import com.kickstarter.libs.utils.WebUtils;
import com.kickstarter.services.KSUri;

import java.io.IOException;

import androidx.annotation.NonNull;
import okhttp3.HttpUrl;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;

public final class ApiRequestInterceptor implements Interceptor {
  private final String clientId;
  private final CurrentUserType currentUser;
  private final String endpoint;
  private final PerimeterXClientType pxManager;
  private final Build build;

  public ApiRequestInterceptor(final @NonNull String clientId, final @NonNull CurrentUserType currentUser,
                               final @NonNull String endpoint, final @NonNull PerimeterXClientType manager,
                               final @NonNull Build build) {
    this.clientId = clientId;
    this.currentUser = currentUser;
    this.endpoint = endpoint;
    this.pxManager = manager;
    this.build = build;
  }

  @Override
  public Response intercept(final @NonNull Chain chain) throws IOException {
    final Response response = chain.proceed(request(chain.request()));
    this.pxManager.intercept(response);
    return response;
  }

  private Request request(final @NonNull Request initialRequest) {
    if (!shouldIntercept(initialRequest)) {
      return initialRequest;
    }

    final Request.Builder builder = initialRequest.newBuilder()
            .addHeader("Accept", "application/json")
            .addHeader("Kickstarter-Android-App-UUID", FirebaseInstanceId.getInstance().getId())
            .addHeader("User-Agent", WebUtils.INSTANCE.userAgent(this.build));

    this.pxManager.addHeaderTo(builder);

    return builder
      .url(url(initialRequest.url()))
      .build();
  }

  private HttpUrl url(final @NonNull HttpUrl initialHttpUrl) {
    final HttpUrl.Builder builder = initialHttpUrl.newBuilder()
      .setQueryParameter("client_id", this.clientId);

    this.currentUser.observable()
      .subscribe(user -> builder.setQueryParameter("oauth_token", this.currentUser.getAccessToken()));

    return builder.build();
  }

  private boolean shouldIntercept(final @NonNull Request request) {
    return KSUri.isApiUri(Uri.parse(request.url().toString()), this.endpoint);
  }
}

File: app/src/main/java/com/kickstarter/services/interceptors/ApiRequestInterceptor.kt
package com.kickstarter.services.interceptors

import android.net.Uri
import com.kickstarter.libs.Build
import com.kickstarter.libs.CurrentUserTypeV2
import com.kickstarter.libs.FirebaseHelper
import com.kickstarter.libs.utils.WebUtils.userAgent
import com.kickstarter.libs.utils.extensions.isApiUri
import okhttp3.HttpUrl
import okhttp3.HttpUrl.Builder
import okhttp3.Interceptor
import okhttp3.Interceptor.Chain
import okhttp3.Request
import okhttp3.Response
import java.io.IOException

class ApiRequestInterceptor(
    private val clientId: String,
    private val currentUser: CurrentUserTypeV2,
    private val endpoint: String,
    private val build: Build
) : Interceptor {

    @Throws(IOException::class)

    override fun intercept(chain: Chain): Response {
        val response: Response = chain.proceed(request(chain.request()))
        return response
    }

    private fun request(initialRequest: Request): Request {
        if (!shouldIntercept(initialRequest)) {
            return initialRequest
        }

        val builder: Request.Builder = initialRequest.newBuilder()
            .addHeader("Accept", "application/json")
            .addHeader("Kickstarter-Android-App-UUID", FirebaseHelper.identifier)
            .addHeader("User-Agent", userAgent(build))

        this.currentUser.accessToken?.let { token ->
            if (token.isNotEmpty()) builder.addHeader("X-Auth", "token $token")
        }

        return builder
            .url(url(initialRequest.url))
            .build()
    }

    private fun url(initialHttpUrl: HttpUrl): HttpUrl {
        val builder: Builder = initialHttpUrl.newBuilder()
            .setQueryParameter("client_id", clientId)
        return builder.build()
    }

    private fun shouldIntercept(request: Request): Boolean {
        return Uri.parse(request.url.toString()).isApiUri(endpoint)
    }
}
